### PR TITLE
Client policies : executor for enforcing DPoP

### LINKED ISSUE #25315: Client policies : executor for enforcing DPoP
### Description

Creating an executor for enforcing sender-constrained refresh token in DPoP, the same as the one for enforsing sender-constrained refresh token in OAuth MTLS `HolderOfKeyEnforcerExecutor`.

- The OAuth 2.1 Authorization Framework
https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-09

Its epic issue is https://github.com/keycloak/keycloak/issues/25141 .

### Discussion

_No response_

### Motivation

_No response_

### Details

_No response_

### PR DESCRIPTION
**Pull request by tnorimat**
closes #25315

### CODE REVIEW

File: services/src/main/java/org/keycloak/services/clientpolicy/executor/DPoPBindEnforcerExecutorFactory.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+import java.util.Collections;
+import java.util.List;
+
+import org.keycloak.Config.Scope;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.KeycloakSessionFactory;
+import org.keycloak.provider.ProviderConfigProperty;
+
+public class DPoPBindEnforcerExecutorFactory  implements ClientPolicyExecutorProviderFactory {
```
Comment (mposolda) [inline services/src/main/java/org/keycloak/services/clientpolicy/executor/DPoPBindEnforcerExecutorFactory.java:28]: Should this maybe implement `EnvironmentDependentFactory` and have this supported just when DPoP feature is enabled? This means that no check is needed in `DPoPBindEnforcerExecutor` itself.

**Review by mposolda — COMMENTED** [review]
@tnorimat Approach looks good to me. There is missing tests and docs. I guess that is the reason why you marked this PR as draft?

File: services/src/main/java/org/keycloak/services/clientpolicy/executor/DPoPBindEnforcerExecutorFactory.java (hunk 1 above)
Comment (tnorimat) [reply services/src/main/java/org/keycloak/services/clientpolicy/executor/DPoPBindEnforcerExecutorFactory.java:28]: Yes, I fill revise it as you suggested.

**Comment by tnorimat** [comment]
@mposolda Yes, this PR lacks tests and docs, therefore I sent the PR as Draft-PR as you inferred. I am now working on them.

**Comment by tnorimat** [comment]
@mposolda I added integration test codes and documentation. Could you check the PR?

File: docs/documentation/server_admin/topics/clients/client-policies.adoc [hunk 2]
```diff
@@ -127,6 +127,7 @@ One of several purposes for this executor is to realize the security requirement
 * Enforce checking if a client is the one to which an intent was issued in a use case where an intent is issued before starting an authorization code flow to get an access token like UK OpenBanking
 * Enforce prohibiting implicit and hybrid flow
 * Enforce checking if a PAR request includes necessary parameters included by an authorization request
+* Enforce <<_dpop-bound-tokens,DPoP-binding tokens>> is used
```
Comment (mposolda) [inline docs/documentation/server_admin/topics/clients/client-policies.adoc:130]:
```suggestion
* Enforce <<_dpop-bound-tokens,DPoP-binding tokens>> is used (available when `dpop` feature is enabled)
```

File: services/src/main/java/org/keycloak/services/clientpolicy/executor/DPoPBindEnforcerExecutorFactory.java (hunk 1 above)
Comment (mposolda) [reply services/src/main/java/org/keycloak/services/clientpolicy/executor/DPoPBindEnforcerExecutorFactory.java:28]: Is it possible to address this point?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/policies/ClientPoliciesExecutorTest.java [hunk 3]
```diff
@@ -1541,4 +1569,164 @@ public void testSecureParContentsExecutorWithRequestObject() throws Exception {
         successfulLoginAndLogout(TEST_CLIENT, TEST_CLIENT_SECRET);
 
     }
+
+    @Test
+    public void testDPoPBindEnforcerExecutor() throws Exception {
```
Comment (mposolda) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/policies/ClientPoliciesExecutorTest.java:1574]: Minor comment, but is it possible to add this test into `DPoPTest` ? As `ClientPoliciesTest` is already quite big and can be nice to have all DPoP related things in the separate place as "DPoP" is still preview and not a core functionality (at least yet).

**Review by mposolda — COMMENTED** [review]
@tnorimat Thanks for the updates Takashi. Added few comments inline.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/policies/ClientPoliciesExecutorTest.java (hunk 3 above)
Comment (tnorimat) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/policies/ClientPoliciesExecutorTest.java]: Yes, I moved a test method to `DPoPTest`.

File: docs/documentation/server_admin/topics/clients/client-policies.adoc (hunk 2 above)
Comment (tnorimat) [reply docs/documentation/server_admin/topics/clients/client-policies.adoc]: Yes, I revised the documentation as you suggested.

File: services/src/main/java/org/keycloak/services/clientpolicy/executor/DPoPBindEnforcerExecutorFactory.java (hunk 1 above)
Comment (tnorimat) [reply services/src/main/java/org/keycloak/services/clientpolicy/executor/DPoPBindEnforcerExecutorFactory.java]: Sorry, I had forgot it. I added `EnvironmentDependentFactory ` to `DPoPBindEnforcerExecutorFactory`.

**Comment by tnorimat** [comment]
@mposolda I revised the PR. Could you check it?

**Review by mposolda — APPROVED** [review]
@tnorimat Thanks!

**Pull request merged**